### PR TITLE
Modify Spouses Tutorial

### PR DESCRIPTION
**Pull request by paroma**
- upgraded pandas version 
- added torch text requirement

- reformatted workshop tutorials to code refactor
- downloaded data link includes train/dev/test data, external KB data, and pertained model

### CODE REVIEW

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    "import pandas as pd"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with open('new_dev_data.pkl', 'rb') as f:\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_1_Task_Intro.ipynb]: Mismatch with file names above (which are better)

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    "    dev_data = pickle.load(f)\n",
+    "    dev_labels = pickle.load(f)\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_1_Task_Intro.ipynb]: Unused/redundant (below too(

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+       "9                                 [and, her, sister]  "
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "dev_data[5:10]"
```
Comment (henryre) [inline tutorials/workshop/Workshop_1_Task_Intro.ipynb]: dev_df.head()

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with open('new_dev_data.pkl', 'rb') as f:\n",
+    "    dev_data = pickle.load(f)\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_1_Task_Intro.ipynb]: dev_df is clearer

File: tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pickle\n",
+    "\n",
+    "with open('new_dev_data.pkl', 'rb') as f:\n",
+    "    dev_data = pickle.load(f)\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb]: Same thing here

File: tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+    "## A. Pattern Matching Labeling Functions\n",
+    "\n",
+    "One powerful form of labeling function design is defining sets of keywords or regular expressions that, as a human labeler, you know are correlated with the true label. For example, we could define a dictionary of terms that occur between person names in a candidate. One simple dictionary of terms indicating a true relation could be:\n",
+    "    \n",
+    "    spouses = {'husband', 'wife'}\n",
+    " \n",
+    "We can then write a labeling function that checks for a match with these terms in the text that occurs between person names. To access the text between the person mentions, we can use the `get_between_tokens` preprocessors from the previous notebook.\n",
+    "\n",
+    "    @labeling_function(resources=dict(spouses=spouses), preprocessors=[get_between_tokens])\n",
+    "    def LF_husband_wife(x: DataPoint, spouses: List[str]) -> int:\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb]: Let's avoid typing in the tutorials

File: tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import List\n",
+    "\n",
+    "from snorkel.labeling.apply import PandasLFApplier\n",
+    "from snorkel.labeling.lf import labeling_function\n",
+    "from snorkel.types import DataPoint\n",
+    "\n",
+    "from spouse_preprocessors import get_left_tokens, \\\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb]: Nit: prefer parens closure to line splices

File: tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+    "get_person_last_names, get_person_text\n",
+    "    \n",
+    "\n",
+    "spouses = {'spouse', 'wife', 'husband', 'ex-wife', 'ex-husband'}\n",
+    "@labeling_function(resources=dict(spouses=spouses))\n",
+    "def LF_husband_wife(x: DataPoint, spouses: List[str]) -> int:\n",
+    "    return 1 if len(spouses.intersection(set(x.between_tokens))) > 0 else 0"
```
Comment (henryre) [inline tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb]: Let's do POSITIVE = 1, ABSTAIN = 0, NEGATIVE = -1 somewhere, then use those for all LFs

File: tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb (hunk 6 above)
Comment (henryre) [reply tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb]: These are also incorrectly typed (should be a Set)

File: tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+     "text": [
+      "Dev Set Scores\n",
+      "{'task/data_valid/valid/accuracy': 0.3624161073825503, 'task/data_valid/valid/f1': 0.5320197044334976, 'task/data_valid/valid/precision': 0.3624161073825503, 'task/data_valid/valid/recall': 1.0}\n",
+      "Test Set Scores\n",
+      "{'task/data_test/test/accuracy': 0.28289473684210525, 'task/data_test/test/f1': 0.441025641025641, 'task/data_test/test/precision': 0.28289473684210525, 'task/data_test/test/recall': 1.0}\n"
+     ]
+    }
+   ],
+   "source": [
+    "model.load('./trained_spouse_model')\n",
```
Comment (henryre) [inline tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb]: Are we going to commit this?

Comment (paroma) [reply tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb]: its in the data download link, not committing

File: tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb (hunk 6 above)
Comment (paroma) [reply tutorials/workshop/Workshop_2_Writing_Labeling_Functions.ipynb]: would this be
`def LF_husband_wife(x, spouses) -> int` or without the ` -> int` part too?

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb (hunk 3 above)
Comment (paroma) [reply tutorials/workshop/Workshop_1_Task_Intro.ipynb]: the first few were confusing so I changed it in a commit, ill go back to `.head()`

File: tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+      "Accuracy: 0.450\n",
+      "F1: 0.164\n",
+      "Precision: 0.092\n",
+      "Recall: 0.768\n",
+      "        y=1    y=2   \n",
+      " l=1    146   1443   \n",
+      " l=2    44    1070   \n"
+     ]
+    }
+   ],
```
Comment (ajratner) [inline tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb]: TODO: Check why MV is doing this badly...

File: tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    "                    markers=markers)\n",
+    "        X.append(marked_tokens)\n",
+    "        \n",
+    "    #featurize string tokens tokens\n",
+    "    featurize_X = featurizer.fit_transform(X)\n",
+    "    return featurize_X\n",
+    "\n",
+    "train_X_tensor = convert_to_lstm_input(train_data)\n",
+    "dev_X_tensor = convert_to_lstm_input(dev_data)\n",
+    "test_X_tensor = convert_to_lstm_input(test_data)"
```
Comment (ajratner) [inline tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb]: I get a `snorkel.mtl module not found` error, I think we need an `__init__.py` file?

Comment (ajratner) [reply tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb]: Everything needs __init__'s, including the sub-folders in mtl

File: tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# # Train SimpleModel\n",
+    "# from snorkel.mtl.trainer import Trainer\n",
+    "# trainer = Trainer(progress_bar=True, n_epochs=5)\n",
+    "# trainer.train_model(model, dataloaders)"
```
Comment (ajratner) [inline tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb]: If we're going to report the test score for end model, should do it for label model as well?

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+   "metadata": {},
+   "source": [
+    "### Data Download\n",
+    "\n",
+    "Make sure `train.pkl`, `dev.pkl`, and `test.pkl` are in the `tutorials/workshop/` directory.\n",
+    "\n",
+    "If you do not have these files, run \n",
+    "\n",
+    "    wget https://www.dropbox.com/s/jmrvyaqew4zp9cy/spouse_data.zip\n",
+    "    unzip spouse_data.zip"
```
[the remark this answers is not in the document]
Comment (henryre) [reply tutorials/workshop/Workshop_1_Task_Intro.ipynb]: Agreed, and I'll handle the download via the bootscript

**Comment by ajratner** [comment]
Another note / todo: `LabelModel` appears to find a spurious optimum voting all negative.  This is not surprising given the generic (accuracy-focused) loss function + class imbalance, but we should set up regularization so that this doesn't happen.

That being said, in the shorter run, can just set the seed properly (in the class initialization vs train method I think?)

File: requirements.txt [hunk 14]
```diff
@@ -4,11 +4,12 @@ jupyter==1.0.0
...
-pandas==0.23.4
+pandas==0.24.0
 pyarrow==0.12.0
 scikit-learn==0.20.2
 scipy==1.2.0
 spacy==2.1.3
 tensorboardX==1.6
 torch==1.1.0
+torchtext==0.3.1
 tqdm==4.29.1
```
[the remark this answers is not in the document]
Comment (henryre) [reply requirements.txt]: @ajratner what doesn't it include? does the addr include the tutorial-redux branch? https://pip.pypa.io/en/stable/reference/pip_install/#git

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+      ]
+     },
+     "execution_count": 6,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "get_text_between.set_mode(PreprocessorMode.PANDAS)\n",
+    "get_text_between(candidate).text_between"
```
[the remark this answers is not in the document]
Comment (henryre) [reply tutorials/workshop/Workshop_1_Task_Intro.ipynb]: Correct, sorry I missed this. @paroma not sure if you were doing this to show an example of the output? But agree with @ajratner , probably better to not show so as not to encourage this usage.

Don't have that usage yet, will add to issues

File: tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+      "[1000 epo]: TRAIN:[loss=0.001]\n",
+      "[1500 epo]: TRAIN:[loss=0.001]\n",
+      "[2000 epo]: TRAIN:[loss=0.001]\n",
+      "[2500 epo]: TRAIN:[loss=0.001]\n",
+      "[3000 epo]: TRAIN:[loss=0.001]\n",
+      "[3500 epo]: TRAIN:[loss=0.001]\n",
+      "[4000 epo]: TRAIN:[loss=0.001]\n",
+      "[4500 epo]: TRAIN:[loss=0.001]\n",
+      "[5000 epo]: TRAIN:[loss=0.001]\n",
+      "Finished Training\n"
```
[the remark this answers is not in the document]
Comment (henryre) [reply tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb]: might also be fixed by `torch.set_num_threads(1)`

**Comment by ajratner** [comment]
@paroma TODO: Add little blurb about how we usually break ties to negative class in info extraction- can do by setting `break_ties=2` in `LabelModel.score`- and still gets way lower than label model!

**Comment by ajratner** [comment]
@paroma last todo: For now fix `LabelModel` init seed

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load_ext autoreload\n",
+    "%autoreload 2\n",
+    "%matplotlib inline\n",
+    "\n",
+    "import pickle\n",
+    "import pandas as pd"
```
[the remark this answers is not in the document]
Comment (paroma) [reply tutorials/workshop/Workshop_1_Task_Intro.ipynb]: added to requirements.txt

File: tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb (hunk 12 above)
Comment (paroma) [reply tutorials/workshop/Workshop_4_Discriminative_Model_Training.ipynb]: doing dev for everything

**Comment by ajratner** [comment]
Quick update: here are the conditional LF probs the `LabelModel` learns in the bad solution:
```
array([0.02317079, 0.01397373, 0.01203453, 0.01      , 0.01      ,
       0.06629263, 0.01      , 0.01      , 0.01      ])
```
So it's just the symmetry breaking issue that we expect; just need to fix the check in `LabelModel` for this.

**Comment by ajratner** [comment]
Okay conclusion (after discussion with PV): We seem to have not ported the newest version of the label model, which does a simple check to resolve this final parameter sign symmetry in the correct way.  Will just port the correct version in either later tonight or after workshop, fix seed that breaks it the correct way for now.  Is also a dead simple thing to explain if someone asks ('always a fundamental sign symmetry left in these problems; in our setting we just check and pick the one that assumes LFs are better; just not in the refactor yet, but dead simple check')

File: tutorials/workshop/Workshop_1_Task_Intro.ipynb (hunk 15 above)
[the remark this answers is not in the document]
Comment (paroma) [reply tutorials/workshop/Workshop_1_Task_Intro.ipynb]: * Moved this into the notebook with the LFs to make the intended usage clear

* Added slides in the talk preceding this notebook to clarify why we have preprocessors

File: tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb (hunk 10 above)
Comment (paroma) [reply tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb]: added `break_ties = 2` and added explanation + random tie breaking comparison

**Comment by paroma** [comment]
done going through the comments except https://github.com/HazyResearch/snorkel/pull/1148#discussion_r296497701 
and
https://github.com/HazyResearch/snorkel/pull/1148#discussion_r296496270

File: requirements.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (ajratner) [reply requirements.txt]: Oh I probably made that mistake... :) . Either way only AWS run is important for workshop

File: tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb (hunk 16 above)
[the remark this answers is not in the document]
Comment (ajratner) [reply tutorials/workshop/Workshop_3_Generative_Model_Training.ipynb]: Yeah also not relevant for workshop, just was on my local MacOS

**Comment by ajratner** [comment]
LGTM!  I think ready to merge in!  Last step is just confirm runs on our AWS setup, can do that when merged

**Comment by henryre** [comment]
Merging, apologies
